src/Week10/AParser.hs

```diff
@@ -0,0 +1,193 @@
+module Week10.AParser where
+
+import Control.Applicative
+import Data.Char
+
+------------------ CONTEXT --------------------
+{-
+ we will represent a parser for a value of type a as a
+ function which takes a String represnting the input to be parsed,
+ and succeeds or fails; if it succeeds, it returns the parsed value along
+ with whatever part of the input it did not use.
+-}
+newtype Parser a = Parser { runParser :: String -> Maybe (a, String) }
+
+{-
+ satisfy takes a Char predicate and constructs a parser which succeeds only
+ if it sees a Char that satisfies the predicate (which it then returns).
+ If it encounters a Char that does not satisfy the predicate (or an empty input), it fails.
+-}
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy p = Parser f
+  where
+    f [] = Nothing -- fail on the empty input
+    f (x:xs) -- check if x satisfies the predicate
+             -- if so, return x along with the remainder
+             -- of the input (that is, xs)
+      | p x = Just (x, xs)
+      | otherwise = Nothing -- otherwise, fail
+
+{-
+ Using satisfy, we can also define the parser char, which expects to
+ see exactly a given character and fails otherwise
+
+ For example:
+ *Parser> runParser (satisfy isUpper) "ABC"
+ Just (’A’,"BC")
+
+ *Parser> runParser (satisfy isUpper) "abc"
+ Nothing
+
+ *Parser> runParser (char ’x’) "xyz"
+ Just (’x’,"yz")
+-}
+char :: Char -> Parser Char
+char c = satisfy (== c)
+
+
+posInt :: Parser Integer
+posInt = Parser f
+  where
+    f xs
+      | null ns = Nothing
+      | otherwise = Just (read ns, rest)
+      where (ns, rest) = span isDigit xs
+
+
+first :: (a -> b) -> (a,c) -> (b,c)
+first f (a, b) = (f a, b)
+--------------- END OF CONTEXT ------------------
+
+{-
+ Exercise 1:
+
+ First, you’ll need to implement a Functor instance for Parser.
+ Hint: You may find it useful to implement a function
+    first :: (a -> b) -> (a,c) -> (b,c)
+-}
+instance Functor Parser where
+  fmap g (Parser f) = Parser $ fmap (first g) . f
+
+
+{-
+ Exercise 2:
+ Now implement an Applicative instance for Parser:
+ • pure a represents the parser which consumes no input and successfully returns a result of a.
+ • p1 <*> p2 represents the parser which first runs p1 (which will
+ consume some input and produce a function), then passes the
+ remaining input to p2 (which consumes more input and produces
+ some value), then returns the result of applying the function to the
+ value. However, if either p1 or p2 fails then the whole thing should
+ also fail (put another way, p1 <*> p2 only succeeds if both p1 and
+ p2 succeed).
+-}
+instance Applicative Parser where
+  pure a = Parser $ \x -> Just (a, x)
+  p1 <*> p2 = Parser p
```

Comment:
I think it's really good for understanding to try to implement this without any existing combinators, so you can see there's actually no magic going on behind the scenes.

```haskell
instance Applicative Parser where
  pure a = Parser $ \x -> Just (a, x)
  Parser f1 <*> Parser f2 = Parser $ \input -> case f1 input of
      Nothing -> Nothing
      Just (f, intermediateOutput) -> case f2 intermediateOutput of
          Nothing               -> Nothing
          Just (a, finalOutput) -> Just (f a, finalOutput)
```

Reply:
This was actually my intuition but I just happened to come across `>>=` in lecture 8 and thought is apt for this particular exercise. But yes,